docs/scalars_and_keras.ipynb

```diff
@@ -440,28 +425,374 @@
+        "    x = self.model(x)\n",
+        "    return x\n",
+        "\n",
+        "# Adds custom batch-level metrics to our previous Functional API model\n",
+        "model = MyModel(create_model())\n",
+        "model.compile(optimizer='adam',\n",
+        "              loss='sparse_categorical_crossentropy',\n",
+        "              metrics=['accuracy'])"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "IlqPZ-h_1_0z"
+      },
+      "source": [
+        "Setting up a summary writer to a different log directory:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "Zjv_7dRwiyf9"
+      },
+      "outputs": [],
+      "source": [
+        "log_dir = 'logs/batch_level/' + datetime.now().strftime(\"%Y%m%d-%H%M%S\") + '/train'\n",
+        "train_writer = tf.summary.create_file_writer(log_dir)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "8wqqNaXP2bLc"
+      },
+      "source": [
+        "Define our Tensorboard callback to log both epoch-level and batch-level metrics to our log directory and call `model.fit()` with our selected `batch_size`: "
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "YXK-iE0e2UOE"
+      },
+      "outputs": [],
+      "source": [
+        "tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir)\n",
+        "\n",
+        "model.fit(x=x_train, \n",
+        "          y=y_train,\n",
+        "          epochs=5,\n",
+        "          batch_size=500, \n",
+        "          validation_data=(x_test, y_test), \n",
+        "          callbacks=[tensorboard_callback])"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "SsxwF83h2kiM"
+      },
+      "source": [
+        "Open Tensorboard with the new log directory and see both the epoch-level and batch-level metrics:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "XlcafPNY2oUW"
+      },
+      "outputs": [],
+      "source": [
+        "%tensorboard --logdir logs/batch_level"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "ReO1LD-g2vgk"
+      },
+      "source": [
+        "### Cumulative batch-level logging"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "T0eE_yjt45TN"
+      },
+      "source": [
+        "Batch-level logging can also be implemented cumulatively, averaging each batch's metrics with those of previous batches in the same epoch and resulting in a smoother training curve when logging batch-level metrics.\n",
+        "\n",
+        "NOTE: Cumulative batch-level logging can show large discontinuities at epoch boundaries due to high batch variance in accuracy and loss."
```

Comment:
Running the notebook, this doesn't actually seem to happen - the curves are completely smooth across epochs. I think that's happening because the metrics aren't being reset on epoch boundaries. The reset behavior would happen automatically if they were added to the model explicitly (via `layer.add_metric()`), but as written now they're just free-floating metric objects - I think we probably want to add them so that the behavior is more like the built-in metrics.

Reply:
Thanks for pointing this out! I realized this early today. After discussing with my manager about cumulative batch-level logging today, however, we actually decided to keep this as-is on purpose, as it's a more useful use case to increase granularity in the metric curve by accumulating across epoch boundaries than to have sudden discontinuities between epochs. Those who want batch-by-batch behavior within the epoch can refer to the instantaneous case. As such, I've edited the description and removed the note on discontinuities.

Reply:
Hmm. I still have significant concerns about showing cumulative averaging across all training steps - I'd like to discuss further in a follow-up. Not resetting at epoch boundaries (and using a non-decaying average) means that any initial outlier values at the start of training will bias the rest of the curve indefinitely, which could be very misleading. For example, see this screenshot:

![image](https://user-images.githubusercontent.com/710113/192877103-ef65d1f3-5c34-4461-bd8c-8898570bca2d.png)

By the end of 5 epochs, the `batch_level` instantaneous loss has pretty much plateaued around 0.1, but the `batch_avg` is still up at 0.2, even though the epoch-level average loss has been below 0.2 since the second epoch (as shown in the logging that `model.fit()` produces or the built-in Keras summaries that write epoch-level loss values).  If all the user wants is smoothing of the batch-level variation, TensorBoard's smoothing algorithm can provide that with greater accuracy, and in a way that can always be adjusted or disabled later to see the true data.

I'll approve to get this PR unblocked, but if we keep this behavior, I'd prefer that you at least reinstate some kind of warning or note in the text, to ensure users realize that this method of averaging will equally weight loss from all batches seen during training, and hence won't match the built-in per-epoch loss metric.